src/svc.rs

```diff
@@ -12,61 +11,120 @@ pub use tower::{service_fn as mk, MakeConnection, MakeService, Service, ServiceE
+    pub fn push_pending(self) -> Layers<Pair<L, pending::Layer>> {
+        self.push(pending::layer())
     }
 
     /// Buffer requests when when the next layer is out of capacity.
-    pub fn buffer_pending<D, Req>(
+    pub fn push_buffer_pending<D, Req>(
         self,
         bound: usize,
         d: D,
-    ) -> Builder<Stack<pending::Layer, Stack<buffer::Layer<D, Req>, L>>>
+    ) -> Layers<Pair<Pair<L, pending::Layer>, buffer::Layer<D, Req>>>
     where
         D: buffer::Deadline<Req>,
         Req: Send + 'static,
     {
-        self.layer(buffer::layer(bound, d)).pending()
+        self.push_pending().push(buffer::layer(bound, d))
+    }
+
+    pub fn push_spawn_ready(self) -> Layers<Pair<L, SpawnReadyLayer>> {
+        self.push(SpawnReadyLayer::new())
+    }
+}
+
+impl<M, L: Layer<M>> Layer<M> for Layers<L> {
+    type Service = L::Service;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        self.0.layer(inner)
+    }
+}
+
+impl<S> Stack<S> {
+    pub fn push<L: Layer<S>>(self, layer: L) -> Stack<L::Service> {
+        Stack(layer.layer(self.0))
     }
 
     /// Buffer requests when when the next layer is out of capacity.
-    pub fn spawn_ready(self) -> Builder<Stack<SpawnReadyLayer, L>> {
-        self.layer(SpawnReadyLayer::new())
+    pub fn push_pending(self) -> Stack<pending::MakePending<S>> {
+        self.push(pending::layer())
     }
 
-    pub fn concurrency_limit(self, max: usize) -> Builder<Stack<ConcurrencyLimitLayer, L>> {
-        Builder(self.0.concurrency_limit(max))
+    /// Buffer requests when when the next layer is out of capacity.
+    pub fn push_buffer_pending<D, Req>(
+        self,
+        bound: usize,
+        d: D,
+    ) -> Stack<buffer::Make<pending::MakePending<S>, D, Req>>
+    where
+        D: buffer::Deadline<Req>,
+        Req: Send + 'static,
+    {
+        self.push_pending().push(buffer::layer(bound, d))
     }
 
-    pub fn load_shed(self) -> Builder<Stack<LoadShedLayer, L>> {
-        Builder(self.0.load_shed())
+    // pub fn push_spawn_ready(self) -> Stack<tower_spawn_ready::MakeSpawnReady<S>> {
+    //     self.push(SpawnReadyLayer::new())
+    // }
+
+    pub fn push_concurrency_limit(self, max: usize) -> Stack<tower::limit::ConcurrencyLimit<S>> {
+        self.push(ConcurrencyLimitLayer::new(max))
     }
 
-    pub fn timeout(self, timeout: Duration) -> Builder<Stack<TimeoutLayer, L>> {
-        Builder(self.0.timeout(timeout))
+    pub fn push_load_shed(self) -> Stack<tower::load_shed::LoadShed<S>> {
+        self.push(LoadShedLayer::new())
     }
 
-    pub fn into_inner(self) -> L {
-        self.0.into_inner()
+    pub fn push_timeout(self, timeout: Duration) -> Stack<tower::timeout::Timeout<S>> {
+        self.push(TimeoutLayer::new(timeout))
     }
 
-    /// Wrap the service `S` with the layers.
-    pub fn service<S>(self, service: S) -> L::Service
+    /// Validates that this stack serves T-typed targets.
+    pub fn serves<T>(self) -> Self
```

Comment:
What was the reason for adding this? It doesn't look like it is used anywhere.

Reply:
As the comment says, this is useful for validating the state of a stack -- i.e. that it actually implements a service for a given type. These assertions aren't necessarily made by the layer implementations, so this is useful during development... We allow(dead_code) here, more generally, because this stuff is here to be useful as you're moving the stack around. It's okay if there are things that aren't actively used...

Reply:
Cool makes sense--thanks!